File: backend/src/recipe.ts
import {AiError} from './core';
import {irHash,SourceIR} from './source-ir';

export const RECIPE_VERSION='processing-recipe-v1';
export type ParseRecipe={recipeVersion:typeof RECIPE_VERSION;stage:'SOURCE_ONLY';parserVersion:string;sourceIRVersion:string;normalizerVersion:string;recipeHash:string};
const isObject=(v:unknown):v is Record<string,unknown>=>!!v&&typeof v==='object'&&!Array.isArray(v);
/** Recomputes the recipe hash with the same canonical JSON as source IR; Python and TypeScript must agree. */
function hashed(recipe:unknown):recipe is Record<string,unknown>&{recipeHash:string} {
  if(!isObject(recipe))return false;
  const {recipeHash,...body}=recipe;
  try {return typeof recipeHash==='string'&&/^[0-9a-f]{64}$/.test(recipeHash)&&body.recipeVersion===RECIPE_VERSION&&irHash(body)===recipeHash;}
  catch {return false;}
}
export function validateParseRecipe(recipe:unknown,ir:SourceIR):asserts recipe is ParseRecipe {
  if(!hashed(recipe)||Object.keys(recipe).sort().join(',')!=='normalizerVersion,parserVersion,recipeHash,recipeVersion,sourceIRVersion,stage'
    ||recipe.stage!=='SOURCE_ONLY'||recipe.parserVersion!==ir.parserVersion||recipe.sourceIRVersion!==ir.schemaVersion||recipe.normalizerVersion!==ir.normalizerVersion)
    throw new AiError('RECIPE_INVALID');
}
/** The extraction recipe must name the stored parse stage it was built on. */
export function validateProcessingRecipe(recipe:unknown,parseRecipeHash:string):string {
  if(!hashed(recipe)||!hashed(recipe.parse)||recipe.parse.recipeHash!==parseRecipeHash||!isObject(recipe.annotation)||!isObject(recipe.index)||typeof recipe.method!=='string')
    throw new AiError('RECIPE_INVALID');
  return recipe.recipeHash;
}
/** Index settings the extraction recipe named for its revision; runs without a recipe (legacy) have none. */
export function expectedIndexSettings(rawJson:unknown):Record<string,unknown>|null {
  const index=isObject(rawJson)&&isObject(rawJson.processingRecipe)?rawJson.processingRecipe.index:null;
  return isObject(index)?index:null;
}
/** The indexer's report of the settings it used must equal the recipe's, whatever the key order. */
export function confirmIndexSettings(expected:Record<string,unknown>,used:unknown):void {
  let same=false;
  try {same=isObject(used)&&irHash(used)===irHash(expected);}catch {same=false;}
  if(!same)throw new AiError('INDEX_RECIPE_MISMATCH');
}

File: ai-service/medical_ai/recipe.py
"""Processing recipe: every version and setting that can change a stage result, with a canonical hash.

The hash is recomputed by the backend (same canonical JSON as source IR), so the recipe holds only
strings, integers, booleans, null and nested objects: floats would serialize differently in Python/JS.
"""
from .chunking import SPLITTER_VERSION
from .extraction import PROMPT_VERSION, SCHEMA_VERSION
from .laboratory import LAB_PROJECTION_VERSION, LAB_VERSION
from .source_ir import IR_VERSION, NORMALIZER_VERSION, canonical_hash
from .visit import VISIT_PROMPT_VERSION, VISIT_VERSION
from .visit_review import REVIEW_PROMPT_VERSION, REVIEW_VERSION

RECIPE_VERSION = 'processing-recipe-v1'


def _hashed(body: dict) -> dict:
    def check(value):
        if isinstance(value, float):
            raise ValueError('Recipe values must not be floats')
        if isinstance(value, dict):
            for item in value.values():
                check(item)
        if isinstance(value, list):
            for item in value:
                check(item)
    check(body)
    return {**body, 'recipeHash': canonical_hash(body)}


def parse_recipe(parser_version: str) -> dict:
    """Deterministic stage before any model call: parser, source IR and normalizer."""
    return _hashed({'recipeVersion': RECIPE_VERSION, 'stage': 'SOURCE_ONLY', 'parserVersion': parser_version,
                    'sourceIRVersion': IR_VERSION, 'normalizerVersion': NORMALIZER_VERSION})


def _digest(models, name):
    return next((m.get('digest') for m in models if isinstance(m, dict) and m.get('name') in {name, name + ':latest'}), None)


def index_settings(settings, models) -> dict:
    """Settings that shape a document's chunks and vectors; the indexer reports the ones it actually used."""
    return {'chunkerVersion': SPLITTER_VERSION, 'chunkSize': settings.chunk_size, 'chunkOverlap': settings.chunk_overlap,
            'embeddingModel': settings.embedding_model, 'embeddingDigest': _digest(models, settings.embedding_model)}


def processing_recipe(settings, provider, parser_version: str, method: str) -> dict:
    """Full recipe of one extraction run. `method` is the path actually taken for this document:
    'lab' (deterministic rows), 'visit', 'visit-review' or 'legacy' (LLM facts)."""
    models = provider.health().get('models', [])
    deterministic = method == 'lab'
    annotation = {
        'lab': {'annotationVersion': LAB_VERSION, 'projectionVersion': LAB_PROJECTION_VERSION, 'promptVersion': None},
        'visit': {'annotationVersion': VISIT_VERSION, 'projectionVersion': None, 'promptVersion': VISIT_PROMPT_VERSION},
        'visit-review': {'annotationVersion': REVIEW_VERSION, 'projectionVersion': None, 'promptVersion': REVIEW_PROMPT_VERSION},
        'legacy': {'annotationVersion': None, 'projectionVersion': None, 'promptVersion': PROMPT_VERSION},
    }[method]
    options = None if deterministic or not hasattr(provider, 'generation_options') else provider.generation_options('TASK: extraction')
    return _hashed({
        'recipeVersion': RECIPE_VERSION,
        'profile': settings.extraction_profile,
        'method': method,
        'parse': parse_recipe(parser_version),
        # Flat keys kept for the existing document API and evaluation reports.
        'sourceIRVersion': IR_VERSION, 'normalizerVersion': NORMALIZER_VERSION,
        'model': None if deterministic else settings.llm_model,
        'modelDigest': None if deterministic else _digest(models, settings.llm_model),
        'generationOptions': options,
        'annotation': {**annotation, 'factSchemaVersion': SCHEMA_VERSION,
                       'maxFactsPerBatch': None if method != 'legacy' else settings.extraction_max_facts_per_batch},
        # Expected index settings; the indexer confirms the ones it used in the revision's index manifest.
        'index': index_settings(settings, models),
    })

